File: Dynamic Programming/minimum-path-sum.py
from typing import  List
# 236ms 5.90%
# 30.4MB 5.01%
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if i == 0 and j == 0:
                    continue
                elif i == 0 and j != 0:
                    grid[i][j] += grid[i][j-1]
                elif i != 0 and j == 0:
                    grid[i][j] += grid[i-1][j]
                else:
                    grid[i][j] += min(grid[i-1][j], grid[i][j-1])
        return grid[-1][-1]

a = Solution()
l = [
  [1,3,1],
  [1,5,1],
  [4,2,1]
]
print(a.minPathSum(l))

File: Dynamic Programming/minimum-path-sum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int const INF = 1e9;
        vector<vector<int>> dp(grid.size(), vector<int>(grid[0].size(), INF));
        dp[0][0] = grid[0][0];
        for (int i=0; i<grid.size(); i++){
            for (int j=0; j<grid[0].size(); j++){
                if (i>0) {
                    dp[i][j] = min(dp[i][j], dp[i-1][j] + grid[i][j]);
                }
                if (j>0){
                    dp[i][j] = min(dp[i][j], dp[i][j-1] + grid[i][j]);
                }
            }
        }
        return dp.back().back();
    }
};

int main(){
    Solution sol;
    vector<vector<int>> grid = {{3}};
    cout << sol.minPathSum(grid) << endl;
    return 0;
}